File: src/main/scala/top/theonly/spark/sca/core/cases/secondsort/SecondSortCase.scala
package top.theonly.spark.sca.core.cases.secondsort

import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}

/**
 * 二次排序
 * 对secondsort.txt文件中的每一行进行排序（降序），先对第一个数字排序，再对第二个数字排序
 */
object SecondSortCase {

  def main(args: Array[String]): Unit = {
    val conf: SparkConf = new SparkConf()
    conf.setMaster("local")
    conf.setAppName("SecondSortCase")

    val context: SparkContext = new SparkContext(conf)

    val rdd: RDD[String] = context.textFile("C:\\code\\IdeaWorkspace\\spark-code\\data\\secondsort.txt")

    val mapRDD: RDD[(MySort, String)] = rdd.map(line => {
      val firstNum: Int = line.split(" ")(0).toInt
      val secondNum: Int = line.split(" ")(1).toInt
      Tuple2(new MySort(firstNum, secondNum), line)
    })
    val sortRDD: RDD[(MySort, String)] = mapRDD.sortByKey(false)

    sortRDD.foreach(tp => println(tp._2))


  }

}

/**
 * 自定义tuple2 key
 * @param firstNum
 * @param secondNum
 */
case class MySort(firstNum: Int, secondNum: Int) extends Ordered[MySort] {
  override def compare(that: MySort): Int = {
    if (this.firstNum == that.firstNum) {
      this.secondNum - that.secondNum
    } else {
      this.firstNum - that.firstNum
    }
  }
}

File: src/main/java/top/theonly/spark/jav/core/cases/secondsort/SecondSortCase.java
package top.theonly.spark.jav.core.cases.secondsort;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.PairFunction;
import org.apache.spark.api.java.function.VoidFunction;
import scala.Tuple2;

/**
 * 二次排序
 * 对secondsort.txt文件中的每一行进行排序（降序），先对第一个数字排序，再对第二个数字排序
 * @author theonly
 */
public class SecondSortCase {

    public static void main(String[] args) {
        SparkConf conf = new SparkConf();
        conf.setMaster("local");
        conf.setAppName("SecondSortCase");

        JavaSparkContext context = new JavaSparkContext(conf);

        JavaRDD<String> rdd = context.textFile("C:\\code\\IdeaWorkspace\\spark-code\\data\\secondsort.txt");

        JavaPairRDD<MySort, String> pairRDD = rdd.mapToPair(new PairFunction<String, MySort, String>() {
            @Override
            public Tuple2<MySort, String> call(String line) throws Exception {
                Integer firstNum = Integer.valueOf(line.split(" ")[0]);
                Integer secondNum = Integer.valueOf(line.split(" ")[1]);
                return new Tuple2<>(new MySort(firstNum, secondNum), line);
            }
        });

        JavaPairRDD<MySort, String> sortRDD = pairRDD.sortByKey(false);

        sortRDD.foreach(new VoidFunction<Tuple2<MySort, String>>() {
            @Override
            public void call(Tuple2<MySort, String> tp) throws Exception {
                System.out.println(tp._2);
            }
        });

    }
}
